File: core/src/main/scala/com/graphbrain/eco/nodes/LetFun.scala
package com.graphbrain.eco.nodes

import com.graphbrain.eco.NodeType.NodeType
import com.graphbrain.eco.{Contexts, NodeType}

class LetFun(params: Array[ProgNode], lastTokenPos: Int= -1) extends FunNode(params, lastTokenPos) {
  override val label = "let"

  override def ntype: NodeType = NodeType.Boolean

  override def booleanValue(ctxts: Contexts) = {
    params(0) match {
      case v: VarNode => {
        val p = params(1)
        p.value(ctxts)
        for (c <- ctxts.ctxts) {
          p.ntype match {
            case NodeType.Boolean => c.setBoolean(v.name, c.getRetBoolean(p))
            case NodeType.Number => c.setNumber(v.name, c.getRetNumber(p))
            case NodeType.Words => c.setWords(v.name, c.getRetWords(p))
            case NodeType.String => c.setString(v.name, c.getRetString(p))
            case NodeType.Vertex => c.setVertex(v.name, c.getRetVertex(p))
          }
          c.setRetBoolean(this, value = true)
        }
      }
    }
  }

  override protected def typeError() = error("parameters must be boolean")
}

File: src/main/java/com/graphbrain/eco/nodes/LetFun.java
package com.graphbrain.eco.nodes;

import com.graphbrain.eco.Context;
import com.graphbrain.eco.Contexts;
import com.graphbrain.eco.NodeType;

public class LetFun extends FunNode {

    public LetFun(ProgNode[] params, int lastTokenPos) {
        super(params, lastTokenPos);
    }

    public LetFun(ProgNode[] params) {
        this(params, -1);
    }

    @Override
    public String label(){return "let";}

    @Override
    public NodeType ntype(Context ctxt){return NodeType.Boolean;}

    @Override
    public void eval(Contexts ctxts) {
        // TODO: check if params[0] is a var
        VarNode v = (VarNode)params[0];
        ProgNode p = params[1];

        p.eval(ctxts);

        for (Context c : ctxts.getCtxts()) {
            switch(p.ntype(c)) {
                case Boolean:
                    c.setBoolean(v.getName(), c.getRetBoolean(p));
                    break;
                case Number:
                    c.setNumber(v.getName(), c.getRetNumber(p));
                    break;
                case Words:
                    c.setWords(v.getName(), c.getRetWords(p));
                    break;
                case String:
                    c.setString(v.getName(), c.getRetString(p));
                    break;
                case Vertex:
                    c.setVertex(v.getName(), c.getRetVertex(p));
                    break;
            }

            c.setRetBoolean(this, true);
        }
    }
}